### ISSUE TITLE
SSL - Trust Server

### ISSUE THREAD

**Issue by tranthanhvu** [tranthanhvu/notes#35]
In the process of working, you were maybe facing problems related to SSL. Some issues will require many steps to set up both server-side and client-side. In this article, I want to mention a simple one, configure with SSL using a self-signed certificate, and set it up on the iOS app?

First, Apple has provided a protocol to handle URLAuthenticationChallenge when received. According to the AuthChallengeDisposition and your business, you could custom the returned credential as below. As you see, I checked that uses had accessed the host before changing its credential. If users accessed my website and required a `.serverTrust`, I will return a new credential.
```
class URLSessionDelegateHandler: NSObject {
    
}

extension URLSessionDelegateHandler: URLSessionDelegate {
    public func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        if challenge.protectionSpace.host == Environment.apiHost {
            if let trust = challenge.protectionSpace.serverTrust {
                let urlCendential = URLCredential(trust: trust)
                completionHandler(.useCredential, urlCendential)
                return
            }
        }

        completionHandler(.performDefaultHandling, nil)
    }
}
```
After that, you have to set up a delegate to URL Session to handle its delegate and throw your credential.
```
let session: URLSession
    
    public init() {
        let delegate = URLSessionDelegateHandler()
        session = URLSession(configuration: URLSessionConfiguration.default, delegate: delegate, delegateQueue: nil)
    }
```

You have to do the same thing if you want to trust the SSL on WebView. You can override the below function by extending from WKNavigationDelegate.
```
func webView(_ webView: WKWebView, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        if challenge.protectionSpace.host == Environment.wvHost {
            if let trust = challenge.protectionSpace.serverTrust {
                let urlCendential = URLCredential(trust: trust)
                completionHandler(.useCredential, urlCendential)
                return
            }
        }
        
        completionHandler(.performDefaultHandling, nil)
    }
```

In short, although It's not a common issue, I have to face off. And It can completely be resolved from server-sides. But in some cases, I still have to solve by myself.
